internal/s3client/client.go

```diff
@@ -53,6 +87,35 @@ func (a *awsEndpointResolver) ResolveEndpoint(service, region string) (aws.Endpo
 
 }
 
+type headerMiddleware struct {
+	client      s3.HTTPClient
+	headers     map[string]string
+	respHeaders http.Header
```

Comment:
Nobody is using this, I think we should remove or at least wire it up.

Reply:
We do use this on line 68
https://github.com/jack-w-shaw/juju/blob/JUJU-4960_separate_s3client/internal/s3client/client.go#L68

True, this only makes a difference if `WithHeaders` is provided as an option, which at this time it is not. This will be directly used soon in a future PR. Happy to move it there if you'd like, the reason it's included here is to remove as much boilerplate from what is looking to be a sizeable PR

Reply:
Pushing back to a future PR